File: test/ModularPipelines.UnitTests/ResultsRepositoryTests.cs
using System.Text.Json;
using Microsoft.Extensions.DependencyInjection;
using ModularPipelines.Attributes;
using ModularPipelines.Context;
using ModularPipelines.Engine;
using ModularPipelines.Extensions;
using ModularPipelines.FileSystem;
using ModularPipelines.Models;
using ModularPipelines.Modules;
using ModularPipelines.TestHelpers;
using Status = ModularPipelines.Enums.Status;

namespace ModularPipelines.UnitTests;

public class ResultsRepositoryTests : TestBase
{
    public static readonly Folder Folder = Folder.CreateTemporaryFolder();

    private class JsonResultRepository : IModuleResultRepository
    {
        public async Task SaveResultAsync<T>(Module<T> module, ModuleResult<T> moduleResult, IPipelineHookContext pipelineContext)
        {
            var file = Folder.CreateFile(module.GetType().FullName!);
            await using var fileStream = file.GetStream();
            await JsonSerializer.SerializeAsync(fileStream, moduleResult);
        }

        public async Task<ModuleResult<T>?> GetResultAsync<T>(Module<T> module, IPipelineHookContext pipelineContext)
        {
            var file = Folder.GetFile(module.GetType().FullName!);
            await using var fileStream = file.GetStream();
            return await JsonSerializer.DeserializeAsync<ModuleResult<T>>(fileStream);
        }
    }

    private class Module1 : Module<IDictionary<string, object>?>
    {
        public override async Task<IDictionary<string, object>?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return null;
        }
    }

    [ModularPipelines.Attributes.DependsOn<Module1>]
    private class Module2 : Module<IDictionary<string, object>?>
    {
        public override async Task<IDictionary<string, object>?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return null;
        }
    }

    [Test]
    [TUnit.Core.NotInParallel(nameof(ResultsRepositoryTests), Order = 1)]
    public async Task RunOne()
    {
        var host = await TestPipelineHostBuilder.Create()
            .AddResultsRepository<JsonResultRepository>()
            .AddModule<Module1>()
            .AddModule<Module2>()
            .BuildHostAsync();

        await host.ExecutePipelineAsync();

        var resultRegistry = host.RootServices.GetRequiredService<IModuleResultRegistry>();
        var module1Result = resultRegistry.GetResult(typeof(Module1))!;
        var module2Result = resultRegistry.GetResult(typeof(Module2))!;

        using (Assert.Multiple())
        {
            await Assert.That(module1Result.ModuleStatus).IsEqualTo(Status.Successful);
            await Assert.That(module2Result.ModuleStatus).IsEqualTo(Status.Successful);
        }
    }

    [Test]
    [TUnit.Core.NotInParallel(nameof(ResultsRepositoryTests), Order = 2)]
    public async Task RunTwoFromHistory()
    {
        var host = await TestPipelineHostBuilder.Create()
            .AddResultsRepository<JsonResultRepository>()
            .AddModule<Module1>()
            .AddModule<Module2>()
            .RunCategories("Other")
            .BuildHostAsync();

        await host.ExecutePipelineAsync();

        var resultRegistry = host.RootServices.GetRequiredService<IModuleResultRegistry>();
        var module1Result = resultRegistry.GetResult(typeof(Module1))!;
        var module2Result = resultRegistry.GetResult(typeof(Module2))!;

        using (Assert.Multiple())
        {
            await Assert.That(module1Result.ModuleStatus).IsEqualTo(Status.UsedHistory);
            await Assert.That(module2Result.ModuleStatus).IsEqualTo(Status.UsedHistory);
        }
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Results/ResultsRepositoryTests.fs
namespace ModularPipelines.UnitTests.FSharp.Results

open System.Text.Json
open Microsoft.Extensions.DependencyInjection
open ModularPipelines.Context
open ModularPipelines.Engine
open ModularPipelines.Extensions
open ModularPipelines.FileSystem
open ModularPipelines.Models
open ModularPipelines.Modules
open ModularPipelines.TestHelpers
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core
open ModularPipelines.Enums

[<AutoOpen>]
module private ResultsRepositoryTestsHelpers =
    let sharedFolder = Folder.CreateTemporaryFolder()

type private JsonResultRepository() =
    interface IModuleResultRepository with
        member _.IsEnabled = true

        member _.SaveResultAsync(m, moduleResult, _) =
            task {
                let file = sharedFolder.CreateFile(m.GetType().FullName)
                use! fileStream = System.Threading.Tasks.Task.FromResult(file.GetStream())
                do! JsonSerializer.SerializeAsync(fileStream, moduleResult)
            }

        member _.GetResultAsync(m, _) =
            task {
                let file = sharedFolder.GetFile(m.GetType().FullName)
                use! fileStream = System.Threading.Tasks.Task.FromResult(file.GetStream())
                return! JsonSerializer.DeserializeAsync<ModuleResult<'T>>(fileStream)
            }

type private RepoModule1() =
    inherit SimpleTestModule<bool>()
    override _.Result = true

[<ModularPipelines.Attributes.DependsOn(typeof<RepoModule1>)>]
type private RepoModule2() =
    inherit SimpleTestModule<bool>()
    override _.Result = true

[<TUnit.Core.NotInParallel(nameof ResultsRepositoryTests, Order = 1)>]
type ResultsRepositoryTests() =
    inherit TestBase()

    [<Test>]
    [<TUnit.Core.NotInParallel(nameof ResultsRepositoryTests, Order = 1)>]
    member _.RunOne() = async {
        let! host =
            TestPipelineHostBuilder
                .Create()
                .AddResultsRepository<JsonResultRepository>()
                .AddModule<RepoModule1>()
                .AddModule<RepoModule2>()
                .BuildHostAsync()
            |> Async.AwaitTask

        do! (host.ExecutePipelineAsync() |> Async.AwaitTask |> Async.Ignore)

        let resultRegistry = host.RootServices.GetRequiredService<IModuleResultRegistry>()
        let module1Result = resultRegistry.GetResult(typeof<RepoModule1>)
        let module2Result = resultRegistry.GetResult(typeof<RepoModule2>)

        do! check(Assert.That(module1Result.ModuleStatus).IsEqualTo(Status.Successful))
        do! check(Assert.That(module2Result.ModuleStatus).IsEqualTo(Status.Successful))
    }

    [<Test>]
    [<TUnit.Core.NotInParallel(nameof ResultsRepositoryTests, Order = 2)>]
    member _.RunTwoFromHistory() = async {
        let! seedHost =
            TestPipelineHostBuilder
                .Create()
                .AddResultsRepository<JsonResultRepository>()
                .AddModule<RepoModule1>()
                .AddModule<RepoModule2>()
                .BuildHostAsync()
            |> Async.AwaitTask

        do! (seedHost.ExecutePipelineAsync() |> Async.AwaitTask |> Async.Ignore)

        let! host =
            TestPipelineHostBuilder
                .Create()
                .AddResultsRepository<JsonResultRepository>()
                .AddModule<RepoModule1>()
                .AddModule<RepoModule2>()
                .RunCategories("Other")
                .BuildHostAsync()
            |> Async.AwaitTask

        do! (host.ExecutePipelineAsync() |> Async.AwaitTask |> Async.Ignore)

        let resultRegistry = host.RootServices.GetRequiredService<IModuleResultRegistry>()
        let module1Result = resultRegistry.GetResult(typeof<RepoModule1>)
        let module2Result = resultRegistry.GetResult(typeof<RepoModule2>)

        do! check(Assert.That(module1Result.ModuleStatus).IsEqualTo(Status.UsedHistory))
        do! check(Assert.That(module2Result.ModuleStatus).IsEqualTo(Status.UsedHistory))
    }
